WordPress/gutenberg
# Add drop-to-upload support for image block

Closes #1701 

This pull request seeks to enhance the image block to allow image upload by drag-and-drop.

![Drag-and-drop](https://user-images.githubusercontent.com/1779930/28180247-92891a2e-67d2-11e7-85f7-08762350b40a.gif)

__Implementation notes:__

Included is a new DropZone component which is largely a ported copy of the one included [in the Calypso project](https://github.com/Automattic/wp-calypso/tree/master/client/components/drop-zone).

An initial attempt had focused on adding the drop zone to the entire editor canvas, but upon observing proposal at https://github.com/WordPress/gutenberg/issues/1563#issuecomment-312894892 to use insertion point, the branch was later changed to focus on the image block. Some refactoring of notices and block removal lingers after this change in direction.

The changes modify the Dashicon component directly, which is not recommended since the file is generated from a separate repository. See follow-up task for updating Dashicon generated behavior.

__Testing instructions:__

Verify that after inserting a new image block, you can drag-and-drop an image from your computer onto the image placeholder to assign the block to use that image.

__Follow-up Tasks:__

- Better error handling for failed uploads
- More accurate file type supports (either by server-side localization, or REST API enhancements to expose this information)
- Block post saving until upload completes
- Update Dashicon component to support `size` prop
